### PR TITLE
Implement New Fee Handling with Tests

### PR DESCRIPTION
**Pull request by asiniscalchi**
## **Type**
enhancement

___

## **Description**
- Introduced new EVM configuration with custom gas management for the Runtime.
- Implemented new fee handling and distribution logic, replacing the old system.
- Added tests for fee and tip distribution, including Ethereum-based transaction fees.
- Configured transaction payment with custom fee handling parameters.
- Updated the Runtime and XCM configuration to use the new fee and EVM management modules.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>evm.rs</strong><dd><code>Introduce EVM Configuration with Custom Gas Management</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/configs/evm.rs
<li>Introduced <code>GAS_PER_SECOND</code> and <code>WEIGHT_PER_GAS</code> constants for EVM gas <br>management.<br> <li> Defined <code>BlockGasLimit</code>, <code>PrecompilesValue</code>, <code>WeightPerGas</code>, and <br><code>GasLimitPovSizeRatio</code> parameter types.<br> <li> Implemented <code>pallet_evm::Config</code> for the <code>Runtime</code> with custom <br>configurations including a new <code>CustomFindAuthor</code> struct for authorship <br>management.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-233ffecc798f02bd4f6727663243a2c3635e5eff84791234c48f51bd1754615e">+64/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mod.rs</strong><dd><code>Implement Fee Handling and Distribution Logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/configs/fee/mod.rs
<li>Introduced <code>DealWithFees</code> struct for handling unbalanced fees and tips.<br> <li> Implemented <code>OnUnbalanced</code> trait for <code>DealWithFees</code> to distribute fees to <br>the block author.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-0afdcde7d15fbaec5addf77d93b49a841f1a23a6ac478304ca5f4cba75c84b44">+47/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction_payment.rs</strong><dd><code>Configure Transaction Payment with Custom Fee Handling</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/configs/transaction_payment.rs
<li>Defined <code>TransactionByteFee</code> and <code>OperationalFeeMultiplier</code> parameter <br>types.<br> <li> Implemented <code>pallet_transaction_payment::Config</code> for the <code>Runtime</code> with <br>custom fee handling.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-5799cf18293f0f5a60383b36ed52c36ce5f1e562f320f3c8dd311a15d8ac6615">+21/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>fee.rs</strong><dd><code>Remove Old Fee Handling Logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/fee.rs
- Removed previous fee handling logic.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-694b84603f302bed52c74c3895636ec0612246a0fbc83219109b291ffe8d8ee2">+0/-26</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Update Runtime with New Fee and EVM Configurations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/lib.rs
<li>Removed old EVM and transaction payment configurations.<br> <li> Updated imports and configurations to use new fee and EVM management <br>modules.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-e71ce79c6361a064296c98f1244126573a4ae185adc40f1afca2fc27e5e74c38">+5/-119</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>xcm_config.rs</strong><dd><code>Update Import Path for Fee Module in XCM Configuration</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/xcm_config.rs
- Updated import path for `fee` module to reflect new structure.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-76e024b17a865749ba6832aa74a8a4fb880b5eb4c8714168e4057013a446b0bb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mock.rs</strong><dd><code>Setup Mock Runtime for Fee Distribution Testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/configs/fee/mock.rs
<li>Setup mock runtime for fee distribution testing with <code>Test</code> enum and <br>configurations.<br> <li> Implemented <code>frame_system::Config</code> and <code>pallet_balances::Config</code> for the <br><code>Test</code> mock.<br> <li> Introduced <code>AuthorAccount</code> struct implementing <code>FindAuthor</code> trait for <br>authorship in tests.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-775e4e31c277273cb5ba993fdd8230510d0f2ef42a211ddd98b537b9d043c8cf">+107/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tests.rs</strong><dd><code>Add Tests for Fee and Tip Distribution</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/configs/fee/tests.rs
<li>Added tests for correct fee and tip distribution to the block author.<br> <li> Included a test for distributing Ethereum-based transaction fees.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/414/files#diff-dfd1c803aca3009f326cedde899cece540acdd61330741a3e4a99f3cb447cd4e">+89/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: ownership-chain/runtime/laos/src/configs/fee/mod.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	}
+}
+
+impl<R> OnUnbalanced<pallet_balances::NegativeImbalance<R>> for DealWithFees<R>
+where
+	R: pallet_balances::Config + pallet_authorship::Config,
+{
+	// Handles unbalanced fees and tips from substrate-based transactions
+	fn on_unbalanceds<B>(
```
Comment (dastanfv) [inline ownership-chain/runtime/laos/src/configs/fee/mod.rs:23]: I think there is no need to implement this. we are sending 100% to author, so only `on_nonzero_unbalanced` is enough. similar to `ToAuthor` in `polkadot` impl

File: ownership-chain/runtime/laos/src/configs/fee/mod.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+use frame_support::traits::{tokens::currency::Currency, Imbalance, OnUnbalanced};
+
+pub struct DealWithFees<R>(PhantomData<R>);
+
+impl<R> DealWithFees<R>
+where
+	R: pallet_balances::Config + pallet_authorship::Config,
+{
+	// Distributes the specified imbalance to the author of the block
+	fn to_author(imbalance: pallet_balances::NegativeImbalance<R>) {
```
Comment (dastanfv) [inline ownership-chain/runtime/laos/src/configs/fee/mod.rs:11]: don't think this deserves a separate function taking into account the comment below

File: ownership-chain/runtime/laos/src/configs/fee/mock.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
+use frame_support::{
```
Comment (dastanfv) [inline ownership-chain/runtime/laos/src/configs/fee/mock.rs:1]: I also don't think it's necessary to have a mock runtime for this. imagine we want to write similar tests for other pallet or type:
1. we create another mock runtime to it
2. realize we have too many mock runtimes and try unify both mock runtime
3. continue like this for several pallets and we end up with a mock runtime which is pretty much the same as the actual runtime

So, I suggest to simply use the actual runtime with `new_test_ext()`

File: ownership-chain/runtime/laos/src/configs/fee/mock.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	where
+		I: 'a + IntoIterator<Item = (ConsensusEngineId, &'a [u8])>,
+	{
+		Some(AUTHOR_ID)
+	}
+}
+
+pub(crate) struct ExtBuilder {}
+
+impl Default for ExtBuilder {
```
Comment (dastanfv) [inline ownership-chain/runtime/laos/src/configs/fee/mock.rs:96]: use `#[derive(Default)]` instead

File: ownership-chain/runtime/laos/src/configs/fee/tests.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		assert_eq!(
+			pallet_balances::Pallet::<Test>::free_balance(author),
+			expected_author_balance,
+			"Author did not receive the correct amount from fees and tips"
+		);
+	});
+}
+
+#[test]
+fn distribute_ethereum_based_fees_correctly() {
```
Comment (dastanfv) [inline ownership-chain/runtime/laos/src/configs/fee/tests.rs:66]: I don't see any difference between this test and `distribute_fees_correctly` (except the comment that it's a "mock ethereum fees")

**Review by dastanfv — CHANGES_REQUESTED** [review]
Overall, IMO this change introduces unnecessary code bloat. modularity might have value, but the functionality/feature here is simple enough that the additional files and overhead don't seem justified.

File: ownership-chain/runtime/laos/src/configs/fee/mod.rs (hunk 1 above)
Comment (asiniscalchi) [reply ownership-chain/runtime/laos/src/configs/fee/mod.rs:23]: very good point

**Pull request closed without merging**